### ISSUE TITLE
Using a Protocol for each  view model

### ISSUE THREAD

**Issue by hilalalhakani** [DeluxeAlonso/UpcomingMovies#345]
Hello 
Actually it's not issue , the code architecture is amazing but i have one question
Why do you use a protocol for each view model .. 
 is it only used to mock the view model  in UnitTest or does it have a different usage ?

**Comment by DeluxeAlonso** [comment]
Hi @hilalalhakani, thanks for your question! One of the benefits of having a protocol for each view model is that we can comply with the Dependency Inversion principle and Open/Closed principle of [SOLID](https://blog.cleancoder.com/uncle-bob/2020/10/18/Solid-Relevance.html).

View models are a dependency in our view controllers. When working with protocols(an abstraction) and not with the concrete class/struct we are making our view controller dependent on abstractions and not on concrete types which is the core of dependency inversion principle. This also facilitates the mocking of these abstractions as you just mentioned and is very helpful when writing our unit tests.

Regarding Open/Closed principle, we are making our classes open for extension and closed for modifications. View controllers do not care about the implementation detail of view models. Lets say we have this protocol and two view model that conforms to it:

```swift
protocol ViewModelProtocol {

    func getAllMovies()

}

struct RealmViewModel: ViewModelProtocol {

    func getAllMovies() {
        // Fetch them locally using realm
    }

}

struct CoreDataViewModel: ViewModelProtocol {

    func getAllMovies() {
        // Fetch them locally using core data
    }

}
```

We can modify/extend how we fetch or which framework to use without changing our View controller or our concrete view models implementations. We just need to **inject** the specific view model we want to use and thats it. If in the future we want to use a different framework we can just create a new view model, make it conform to ViewModelProtocol and inject it on the view controller.

In conclusion, it makes our code more scalable and flexible and also facilitates dependency injection a lot.
Hope this helps!  If you have any other questions don't hesitate to ask.

**Comment by hilalalhakani** [comment]
Hello Deluxe ,
Thank you for your response 
 
 yes you're right about that but  can we do it in the localDataSource instead of doing it in the viewModel 
 for example `MovieSearchLocalDataSourceProtocol` that contains
 
```
 public protocol MovieSearchLocalDataSourceProtocol {
    var didUpdateMovieSearch: (() -> Void)? { get set }
    func getMovieSearches() -> [MovieSearch]
    func save(with searchText: String)
}
```
so the code can be changed from 
  ```
func getMovieSearches() {
        // Fetch them locally using core data
    }
```
to 

  ```
func getMovieSearches() {
        // Fetch them locally using Realm
   }
```
in this case only the  **LocalDataSourceProtocol** and  **RemoteDataSourceProtocol**  for each useCase are conforming to protocols
but not the view model since there would be no changes to it ..

**Comment by DeluxeAlonso** [comment]
Hey @hilalalhakani, yes you are right, thats how it is done in the app. I was using the realm/core data just as an example :sweat_smile:. 

Maybe a better example is the `MovieCreditsViewController`.  Lets say that I want to show in some places only the cast and not the crew anymore.
We could support both features without modifying MovieVideosViewController. We would only need to create something like this:

```
class OnlyCastCreditsViewModel: MovieCreditsViewModelProtocol {
  func getMovieCredits() {
    // This is implementation detail, here we can only retrieve cast people and not crew.
  }
}
```

And then inject it as a dependency to an instance of `MovieCreditsViewController`. We can easily do this because `MovieCreditsViewController` depends on `MovieCreditsViewModelProtocol`(abstraction) and not on `MovieCreditsViewModel`(concrete type)

The view controller is not modified at all and also does not have any knowledge of the implementation detail of the view model.  `MovieCreditsViewController` does not have to take any decision, it only presents whatever the view model is communicating. This give us a lot of flexibility.

I hope this was a better example. And again, dont hesitate to ask any question!

**Comment by hilalalhakani** [comment]
Thanks a lot for your help Deluxe ,
now i understand , you're using the same view for many viewModels 
Great job !!
